### PR TITLE
Feat/focal loss faster rcnn

### PR DESCRIPTION
**Pull request by muaz-urwa**
This pull request adds focal loss to faster rcnn classification loss computation. Currently cross entropy loss is used to for classification:

CE(pt) =−log(pt)
where pt = softmax probability of correct class

Focal loss as defined [here](https://arxiv.org/pdf/1708.02002.pdf) is:
FL(pt) =−(1−pt)^γ log(pt)

Alpha balancing version of focal loss is:
FL(pt) =−αt(1−pt)^γ log(pt)

We are not using the alpha balancing. Alpha balancing is used to tackle imbalance between foreground and background. Since the two stage network intrinsically handles this, we can use foreground background sampling parameters in faster rcnn to handle this kind of imbalance. But imbalance between foreground object classes is not handled, I have used the version of focal loss without alpha balancing for it. 

Focal loss also has an effect similar to hard negative mining, by reducing the contribution of easy examples to loss. Since we don't have hard negative miner in faster rcnn currently, focal loss becomes handy.

I observed slight improvements in my model performance in imbalanced class problems. Interestingly, it is important to make sure all(most) annotations are correct, since I observed that in cases you have some annotation mistakes focal loss can hurt you by over emphasizing on annotation mistakes.

### CODE REVIEW

**Comment by muaz-urwa** [comment]
Will resolve the linting errors soon.

**Comment by muaz-urwa** [comment]
Need help with python_type_check failure:

"torchvision/models/detection/roi_heads.py:48: error: Type signature has too few arguments  [syntax]
    def fastrcnn_loss(class_logits, box_regression, labels, regression_tar"

**Comment by muaz-urwa** [comment]
@fmassa 

Thank you for your feedback.

1) it only gets exposed to Faster R-CNN:
That makes total sense, if focal loss is being integrated in pytorch there should be a single point of definition that can be reused over modules.

2) the alpha coefficient is not exposed, and seems a bit different than the original implementation
That was on purpose, because in my understanding focal loss with alpha balancing is designed for single stage detectors so they can handle imbalance between foreground and background. But two stage detectors solve this problem by passing on limited proposals and then using sampling to use fixed proportion of background and foreground examples. Tuning following hyper parameters in faster rcnn adresses this issue: _box_batch_size_per_image=512, box_positive_fraction=0.25_
This focal loss is designed to handle imbalance between foreground classes. But again, it can be argued that this focal loss is a special case of focal loss with alpha balancing with _alpha = 0.5_. So single definition of focal loss makes more sense.


3) Maybe this could be achieved by having an attribute _self.box_loss_ that defaults to _fastrcnn_loss_
I think that is great idea. Rather than making this _self.box_loss_ a function pointer we can pass in a class (nn.Module) with a fixed signature for forward call. So other arguments (like gama, alpha etc) can be set during initialization. If the user does not specify this _self.box_loss_ class, we can set it to _fastrcnn_loss_ class in faster rcnn constructor. Similar design has been used for _box_roi_pool, box_head, box_predictor_. They can be passed by user, otherwise they are set in the constructor to default classes.

What are your thoughts about this ?

**Comment by fmassa** [comment]
Hi @muaz-urwa 

So, I think we should try to make the implementation of Faster R-CNN in torchvision as simple to use and modify as possible.

Given that adding focal loss to Faster R-CNN is not something standard (at least I haven't seen it being widely applied), for this particular implementation I'm tempted to say that it would be preferable not to merge it to torchvision.

Now, I think it's still up for discussion to allow to customize the `loss` function for boxes / masks / keypoints, but as I said before, this is a simplified customization because we restrict what can be visible to each loss arguments, so in some sense it's not really as generic as possible.

Let me know what you think

**Comment by muaz-urwa** [comment]
@fmassa 

Yes I totally understand that merging this PR does not make sense. Sorry for not being clear in my previous comment. 

In point 1 and 2, I was agreeing with you that this PR should be discarded .

In point 3, I was only discussing the alternative that you mentioned in your first comment:

"_I do think though that we should provide an easier way to customize the losses for each one of the heads.
Maybe this could be achieved by having an attribute self.box_loss that defaults to fastrcnn_loss. But the problem with this is that in some way it adds complexity but is not as flexible as it might first look -- indeed, the input signature is fixed, and we can't pass other things to the loss other than class_logits, box_regression, labels, regression_targets._"

**Comment by fmassa** [comment]
Hi @muaz-urwa 

Thanks for clarifying your points, and sorry if I misunderstood something.

About point 3, I still think it's worth discussing if the fact that the input signature to the loss being fixed would be a severe limitation or not.

**Comment by muaz-urwa** [comment]
@fmassa no problem. Thank you for your informative insights :)
